microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MicroProfileMetricsSubsystemRemove.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.microprofile.metrics;
+
+import org.jboss.as.controller.AbstractRemoveStepHandler;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.dmr.ModelNode;
+
+/**
+ * @author <a href="http://jmesnil.net/">Jeff Mesnil</a> (c) 2018 Red Hat inc.
+ */
+public class MicroProfileMetricsSubsystemRemove extends AbstractRemoveStepHandler {
+
+    @Override
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        super.performRuntime(context, operation, model);
+
+        context.removeService(MetricsContextService.SERVICE_NAME);
+        context.removeService(MicroProfileMetricsSubsystemDefinition.WILDFLY_REGISTRATION_SERVICE);
+    }
```

Comment:
I think ServiceRemoveStepHandler can do this.

Tangent: it might be nicer if ServiceRemoveStepHandler didn't need to be told what capabilities to remove in its constructor and could be set to just see what's registered with the MRR.

Reply:
I think I have fixed it but the `ServiceRemoveStepHandler(org.jboss.as.controller.AbstractAddStepHandler, org.jboss.as.controller.capability.RuntimeCapability...)` constructor I used is deprecated. I'm not sure why it is deprecated as my use case seems valid (2 capabilities are associated to the resource).

Reply:
@jmesnil The deprecated probably relates to my point about "it might be nicer if ServiceRemoveStepHandler didn't need to be told what capabilities to remove in its constructor and could be set to just see what's registered with the MRR."  A lot of the core handlers took RuntimeCapability... as a constructor param and used the data to record the capability. But then they were changed to get the capabilities from the MRR, making adding via the constructor an out-of-date code style. So all those constructor variants got deprecated. But ServiceRemoveStepHandler uses the capabilities for more than just recording them. So the deprecation there was premature.

Reply:
Oh, I see that my "it might be nicer" comment is wrong. Looking more closely ServiceRemoveStepHandler already does what I want. It gets the capabilities from the MRR if they were not provided via the constructor param.

So, I'm not sure what to say about the deprecation, other than perhaps this c'tor should be public now:

```
protected ServiceRemoveStepHandler(final AbstractAddStepHandler addOperation)
```